lib/datadog/appsec/utils/read_write_lock.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Datadog
+  module AppSec
+    module Utils
+      # Simple implentation for a fair read write lock
+      # It allows for as many concurrent readers but only one concurrent writer.
+      # To avoid the writer thread being exhausted, once a writer thread has signaled that it wants to acquire the lock,
+      # no more reader threads can acquire the lock.
+      # The new reader threads will wait for the writer thread to finish.
+      class ReadWriteLock
+        def initialize
+          @reader_mutex    = Mutex.new
+          @reader_q        = ConditionVariable.new
+          @reader_count    = 0
+          @reader_releases = 0
+
+          @writer_mutex    = Mutex.new
+          @writer_q        = ConditionVariable.new
+          @writter         = false
+        end
+
+        def with_rlock
+          rlock
+          yield
+        ensure
+          runlock
+        end
+
+        def with_lock
+          lock
+          yield
+        ensure
+          unlock
+        end
+
+        def rlock
+          @reader_mutex.synchronize do
+            @reader_count += 1
+
+            writter = false
+
+            @writer_mutex.synchronize do
+              writter = @writter
+            end
+
+            @reader_q.wait(@reader_mutex) if writter
```

Comment:
IIUC the `@writer_mutex.synchronize` makes `rlock` wait once a write lock attempt has started (signalled by the same `@writer_mutex.synchronize` in `#lock`) in order to implement this:

```
      # To avoid the writer thread being exhausted, once a writer thread has signaled that it wants to acquire the lock,
      # no more reader threads can acquire the lock.
```

Then `@writer_q.wait(@writer_mutex) while (@reader_releases != @reader_count)` makes `lock` wait until all readers have released their locks.

The ` || @writter` part is what makes it wait if another writer is in progress, so it's what implements the lock for writers.

Right after that `lock` sets `@writter = true` to actually acquire the (write) lock.

Since `@writter` is a boolean this is what makes it have only a single writer as described:

```
      # It allows for ... only one concurrent writer.
```

Now, back to this highlight, the intent of `@reader_q.wait(@reader_mutex) if writter` seems to be that if the writer lock is held by a writer then we wait for it to be unlocked to proceed with the remainder of `rlock` (which happens to be nothing), which effectively blocks the read lock to be acquired until the write lock is released.

`writter` is obtained from `@writter` within the `@writer_mutex.synchronize` block, and used after the end of that block. This gives the opportunity for the Ruby VM to switch threads, so we must ensure that in all cases `writter` and `@writter` keep having the same value, otherwise `if writter` will not check reality.

After the entirety of either `lock` or `unlock` there are two cases.

If it is after `unlock` then `writter` will be `false`. By that point a writer cannot alter `@writter` to `true` (which is only done via `lock`) because of a) `@reader_count += 1` above and b) no other reader can `runlock` in a way that would compensate that `+= 1` (because of the surrounding `@reader_mutex.synchronize`). So it seems like we're good here.

If it is after `lock` then `writter` will be `true`. By that point a writer has locked the lock, and in between `synchronize do .. end` and `@reader_q.wait(@reader_mutex) if writter` that writer could alter `@writter` to `false` (which, by the way, is only done via `unlock`).

Even though `@writter` is now `false`, `writter is still `true` and `@reader_q.wait(@reader_mutex)` will be called. After having set `@writter`, `unlock` will attempt  `@reader_mutex.synchronize { @reader_q.broadcast }` but be unable to proceed since `@reader_mutex` is held by `rlock`.

So on each side we have:

- `@reader_mutex.synchronize do ... @reader_q.wait(@reader_mutex) ... end` waiting for `@reader_q.broadcast` to be unblocked
- `@reader_mutex.synchronize { @reader_q.broadcast }` waiting for `@reader_mutex.synchronize do ... end` to be unblocked

Which if I'm not mistaken would be a deadlock.

Reply:
@lloeki, your explanation makes sense; the code could lead to a deadlock situation.